### ISSUE TITLE
Euterpea2 is incompatible with the current PortMidi version

### ISSUE THREAD

**Issue by sternenseemann** [Euterpea/Euterpea2#23]
Because of its dependency on an old version of PortMidi this currently happens, when installing:

```
compileBuildDriverPhase
setupCompileFlags: -package-db=/build/setup-package.conf.d -j1 -threaded
[1 of 1] Compiling Main             ( Setup.hs, /build/Main.o )
Linking Setup ...
configuring
configureFlags: --verbose --prefix=/nix/store/vqxr1zdzgvx9lwd4acbqjkj58pc26aiy-Euterpea-2.0.5 --libdir=$prefix/lib/$compiler --libsubdir=$pkgid --docdir=/nix/store/5bfbip1w8ldcqqlj70mxjjrfigk7b8ld-Euterpea-2.0.5-doc/share/doc/Euterpea-2.0.5 --with-gcc=gcc --package-db=/build/package.conf.d --ghc-option=-j1 --disable-split-objs --enable-library-profiling --profiling-detail=all-functions --disable-profiling --enable-shared --disable-coverage --enable-static --disable-executable-dynamic --enable-tests --enable-library-vanilla --enable-library-for-ghci --ghc-option=-split-sections --extra-lib-dirs=/nix/store/w8bxy3xrvrd1ci5772y7arkicj0q7x3h-ncurses-6.1/lib --extra-lib-dirs=/nix/store/d60s5bwv6r3hl6w8cg5j7znwaiva43n8-gmp-6.1.2/lib
Using Parsec parser
Configuring Euterpea-2.0.5...
CallStack (from HasCallStack):
  die', called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:958:20 in Cabal-2.2.0.1:Distribution.Simple.Configure
  configureFinalizedPackage, called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:462:12 in Cabal-2.2.0.1:Distribution.Simple.Configure
  configure, called at libraries/Cabal/Cabal/Distribution/Simple.hs:596:20 in Cabal-2.2.0.1:Distribution.Simple
  confHook, called at libraries/Cabal/Cabal/Distribution/Simple/UserHooks.hs:67:5 in Cabal-2.2.0.1:Distribution.Simple.UserHooks
  configureAction, called at libraries/Cabal/Cabal/Distribution/Simple.hs:178:19 in Cabal-2.2.0.1:Distribution.Simple
  defaultMainHelper, called at libraries/Cabal/Cabal/Distribution/Simple.hs:115:27 in Cabal-2.2.0.1:Distribution.Simple
  defaultMain, called at Setup.hs:2:8 in main:Main
Setup: Encountered missing dependencies:
PortMidi ==0.1.6.1
```

I've submitted a fix as #22, which can also be downloaded as a [patch](https://patch-diff.githubusercontent.com/raw/Euterpea/Euterpea2/pull/22.patch).

**Comment by donya** [comment]
Please give a full transcript (what command lead to this?) and OS details. I can't reproduce this with a clean install of Euterpea on my machines, which have no trouble installing the PortMidi 0.1.6.1 dependency.

**Comment by sternenseemann** [comment]
NixOS and hackage have already moved past this version, the current one is [0.2.0.0](https://hackage.haskell.org/package/PortMidi).

**Comment by donya** [comment]
Please give me precise details so I can understand what is happening here. 

1. You say it is "more of less" a standard cabal install. What exactly was run?

2. I would like to know the exact reason why 0.1.6.1 doesn't work, because it is fine on other operating systems including, as far as I know, other Linux distributions. From various e-mail conversations in addition to past issues reported via github, NixOS seems very finicky about requiring more recent library versions than is the case for other operating systems. If this is going to be a persistent problem with it then I need to list it as an OS for which the library can't be guaranteed (not purely because of this particular PortMidi version issue; this is simply the most recent such case where it breaks NixOS but not other OS varieties).

**Comment by sternenseemann** [comment]
Well it is cabal install wrapped by Nix, meaning that it is equivalent to cabal install, but Nix takes care of providing the dependencies.

The problem on NixOS is simply, that they are quick to update their package version to the most recent  version from hackage.

**Comment by donya** [comment]
Thanks for the additional info. Unfortunately, that means NixOS technically falls under the category of "other package managers" on the setup instructions page (there is a compatibility warning about those; I only guarantee support for plain vanilla cabal installs). It is strange to me that NixOS rejects package versions that aren't even that old, but it explains why it has been a problematic OS for Euterpea over time.

The current rule I use in updating dependencies for Euterpea is to focus on updates required for new Haskell Platform versions when they are released, and not bumping dependency versions until required by that or unless some kind of wide-spread other compatibility problem emerges. While it may seem strange to adhere to older versions, it's a strategy has dramatically cut down on the number of Windows/Mac complaints I receive over time and has increased backwards compatibility (many Euterpea users do not have up-to-date OS/GHC versions), so I will not be changing that approach. For quite some time, NixOS has been the only OS to result in complaints other than those associated with new Haskell Platform releases, so I am going to list it as an unsupported system on the website. 

My suggestion is that some NixOS should fork Euterpea and update it as needed to keep it compatible with NixOS if it requires the most recent package versions of Euterpea's dependencies at all times. If someone wants to make a separate forked repo for that, I would be happy to include a link to it from the setup page on euterpea.com so that other NixOS users can find it more easily.

**Comment by basvandijk** [comment]
Note that nixpkgs now [uses](https://github.com/NixOS/nixpkgs/commit/9b67908e6799aeb1665f6424f2c3df5352d22aa0) #22 to fix the build.

**Comment by jwaldmann** [comment]
@sternenseemann  when I clone your repo, and "cabal install" (with ghc-8.6) I get
```
cabal: Could not resolve dependencies:
next goal: Euterpea (user goal)
rejecting: Euterpea-2.0.6 (constraint from user target requires ==2.0.5)
trying: Euterpea-2.0.5
trying: base-4.12.0.0/installed-4.1... (dependency of Euterpea-2.0.5)
next goal: stm (dependency of Euterpea-2.0.5)
rejecting: stm-2.5.0.0/installed-CMx..., stm-2.5.0.0 (conflict: Euterpea =>
stm>=2.4 && <2.5)
...
```
I can fix this with `cabal install --allow-newer`

**Comment by donya** [comment]
This is addressed in changes as part of version 2.0.7 (on GitHub only for now; will go to Hackage after later more extensive testing)
